File: backend/app/schemas/health.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator


ObservationCategory = Literal["chronic-vitals", "lifestyle", "body-vitals"]
ObservationSource = Literal["manual", "device", "ai-extract"]
ClinicalRecordSource = Literal["manual", "ai-extract"]
ConditionCategory = Literal["diagnosis", "chronic", "allergy", "family-history"]
ConditionStatus = Literal["active", "inactive", "resolved"]
MedicationStatus = Literal["active", "stopped"]
EncounterType = Literal["outpatient", "inpatient", "checkup", "emergency"]
CarePlanCategory = Literal[
    "medication-reminder",
    "activity-reminder",
    "checkup-reminder",
    "health-advice",
    "daily-tip",
]
CarePlanIconKey = Literal["medication", "exercise", "checkup", "meal", "rest", "social", "general"]
CarePlanTimeSlot = Literal["清晨", "上午", "午后", "晚间", "睡前"]
CarePlanStatus = Literal["active", "completed", "cancelled"]
CarePlanGeneratedBy = Literal["ai", "manual"]
HealthSummaryStatus = Literal["good", "warning", "alert"]
HealthRecordActionType = Literal["create", "update", "delete"]
HealthRecordActionResource = Literal["observations", "conditions", "medications", "encounters"]


class ObservationBase(BaseModel):
    category: ObservationCategory
    code: str
    display_name: str
    value: float | None = None
    value_string: str | None = None
    unit: str | None = None
    context: str | None = None
    effective_at: str
    source: ObservationSource
    device_name: str | None = None
    notes: str | None = None

    @field_validator("code", "display_name")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class ObservationCreate(ObservationBase):
    pass


class ObservationUpdate(BaseModel):
    category: ObservationCategory | None = None
    code: str | None = None
    display_name: str | None = None
    value: float | None = None
    value_string: str | None = None
    unit: str | None = None
    context: str | None = None
    effective_at: str | None = None
    source: ObservationSource | None = None
    device_name: str | None = None
    notes: str | None = None

    @field_validator("code", "display_name")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class ObservationRead(ObservationBase):
    id: str
    member_id: str
    created_at: str
    updated_at: str


class HealthRecordObservationDraft(BaseModel):
    category: ObservationCategory
    code: str
    display_name: str
    value: float | None = None
    value_string: str | None = None
    unit: str | None = None
    context: str | None = None
    effective_at: str
    notes: str | None = None


class ConditionBase(BaseModel):
    category: ConditionCategory
    display_name: str
    clinical_status: ConditionStatus
    onset_date: str | None = None
    source: ClinicalRecordSource
    notes: str | None = None

    @field_validator("display_name")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class ConditionCreate(ConditionBase):
    pass


class ConditionUpdate(BaseModel):
    category: ConditionCategory | None = None
    display_name: str | None = None
    clinical_status: ConditionStatus | None = None
    onset_date: str | None = None
    source: ClinicalRecordSource | None = None
    notes: str | None = None

    @field_validator("display_name")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class ConditionRead(ConditionBase):
    id: str
    member_id: str
    created_at: str
    updated_at: str


class HealthRecordConditionDraft(BaseModel):
    category: ConditionCategory
    display_name: str
    clinical_status: ConditionStatus
    onset_date: str | None = None
    notes: str | None = None


class MedicationBase(BaseModel):
    name: str
    indication: str | None = None
    dosage_description: str | None = None
    status: MedicationStatus
    start_date: str | None = None
    end_date: str | None = None
    source: ClinicalRecordSource

    @field_validator("name")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Medication name is required.")
        return cleaned


class MedicationCreate(MedicationBase):
    pass


class MedicationUpdate(BaseModel):
    name: str | None = None
    indication: str | None = None
    dosage_description: str | None = None
    status: MedicationStatus | None = None
    start_date: str | None = None
    end_date: str | None = None
    source: ClinicalRecordSource | None = None

    @field_validator("name")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Medication name is required.")
        return cleaned


class MedicationRead(MedicationBase):
    id: str
    member_id: str
    created_at: str
    updated_at: str


class HealthRecordMedicationDraft(BaseModel):
    name: str
    indication: str | None = None
    dosage_description: str | None = None
    status: MedicationStatus
    start_date: str | None = None
    end_date: str | None = None


class EncounterBase(BaseModel):
    type: EncounterType
    facility: str | None = None
    department: str | None = None
    attending_physician: str | None = None
    date: str
    summary: str | None = None
    source: ClinicalRecordSource


class EncounterCreate(EncounterBase):
    pass


class EncounterUpdate(BaseModel):
    type: EncounterType | None = None
    facility: str | None = None
    department: str | None = None
    attending_physician: str | None = None
    date: str | None = None
    summary: str | None = None
    source: ClinicalRecordSource | None = None


class EncounterRead(EncounterBase):
    id: str
    member_id: str
    created_at: str
    updated_at: str


class HealthRecordEncounterDraft(BaseModel):
    type: EncounterType
    facility: str | None = None
    department: str | None = None
    attending_physician: str | None = None
    date: str
    summary: str | None = None


class SleepRecordBase(BaseModel):
    start_at: str
    end_at: str
    total_minutes: int
    deep_minutes: int | None = None
    rem_minutes: int | None = None
    light_minutes: int | None = None
    awake_minutes: int | None = None
    efficiency_score: float | None = None
    is_nap: bool = False
    source: ObservationSource
    device_name: str | None = None


class SleepRecordCreate(SleepRecordBase):
    pass


class SleepRecordUpdate(BaseModel):
    start_at: str | None = None
    end_at: str | None = None
    total_minutes: int | None = None
    deep_minutes: int | None = None
    rem_minutes: int | None = None
    light_minutes: int | None = None
    awake_minutes: int | None = None
    efficiency_score: float | None = None
    is_nap: bool | None = None
    source: ObservationSource | None = None
    device_name: str | None = None


class SleepRecordRead(SleepRecordBase):
    id: str
    member_id: str
    created_at: str


class WorkoutRecordBase(BaseModel):
    type: str
    start_at: str
    end_at: str
    duration_minutes: int
    energy_burned: float | None = None
    distance_meters: float | None = None
    avg_heart_rate: int | None = None
    source: ObservationSource
    device_name: str | None = None
    notes: str | None = None

    @field_validator("type")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Workout type is required.")
        return cleaned


class WorkoutRecordCreate(WorkoutRecordBase):
    pass


class WorkoutRecordUpdate(BaseModel):
    type: str | None = None
    start_at: str | None = None
    end_at: str | None = None
    duration_minutes: int | None = None
    energy_burned: float | None = None
    distance_meters: float | None = None
    avg_heart_rate: int | None = None
    source: ObservationSource | None = None
    device_name: str | None = None
    notes: str | None = None

    @field_validator("type")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Workout type is required.")
        return cleaned


class WorkoutRecordRead(WorkoutRecordBase):
    id: str
    member_id: str
    created_at: str


class HealthSummaryBase(BaseModel):
    category: str
    label: str
    value: str
    status: HealthSummaryStatus
    generated_at: str

    @field_validator("category", "label", "value")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class HealthSummaryCreate(HealthSummaryBase):
    pass


class HealthSummaryUpdate(BaseModel):
    category: str | None = None
    label: str | None = None
    value: str | None = None
    status: HealthSummaryStatus | None = None
    generated_at: str | None = None

    @field_validator("category", "label", "value")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class HealthSummaryRead(HealthSummaryBase):
    id: str
    member_id: str
    created_at: str


class CarePlanBase(BaseModel):
    category: CarePlanCategory
    icon_key: CarePlanIconKey | None = None
    time_slot: CarePlanTimeSlot | None = None
    title: str
    description: str
    notes: str | None = None
    status: CarePlanStatus
    scheduled_at: str | None = None
    completed_at: str | None = None
    generated_by: CarePlanGeneratedBy
    assignee_member_id: str | None = None

    @field_validator("title", "description")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned

    @field_validator("notes", "assignee_member_id")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class CarePlanCreate(CarePlanBase):
    pass


class CarePlanUpdate(BaseModel):
    category: CarePlanCategory | None = None
    icon_key: CarePlanIconKey | None = None
    time_slot: CarePlanTimeSlot | None = None
    title: str | None = None
    description: str | None = None
    notes: str | None = None
    status: CarePlanStatus | None = None
    scheduled_at: str | None = None
    completed_at: str | None = None
    generated_by: CarePlanGeneratedBy | None = None
    assignee_member_id: str | None = None

    @field_validator("title", "description", "notes", "assignee_member_id")
    @classmethod
    def validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned


class CarePlanRead(CarePlanBase):
    id: str
    member_id: str
    created_at: str
    updated_at: str


class ObservationTrendPoint(BaseModel):
    id: str
    effective_at: str
    value: float | None = None
    value_string: str | None = None
    unit: str | None = None
    notes: str | None = None


class ObservationTrendRead(BaseModel):
    member_id: str
    code: str
    display_name: str | None = None
    points: list[ObservationTrendPoint]


class DashboardMember(BaseModel):
    id: str
    name: str
    gender: str
    avatar_url: str | None = None
    blood_type: str | None = None


class DashboardMemberSummary(BaseModel):
    member: DashboardMember
    health_summaries: list[HealthSummaryRead] = Field(default_factory=list)


class DashboardReminder(CarePlanRead):
    member_name: str


class DashboardReminderGroup(BaseModel):
    time_slot: CarePlanTimeSlot
    reminders: list[DashboardReminder] = Field(default_factory=list)


class DashboardRead(BaseModel):
    members: list[DashboardMemberSummary]
    today_reminders: list[DashboardReminder]
    reminder_groups: list[DashboardReminderGroup] = Field(default_factory=list)
    today_reminders_refreshed_at: str | None = None


class DailyGenerationRefreshResult(BaseModel):
    member_ids: list[str] = Field(default_factory=list)
    failed_member_ids: list[str] = Field(default_factory=list)
    errors: dict[str, str] = Field(default_factory=dict)


class DailyGenerationRefreshRequest(BaseModel):
    language: Literal["zh", "en"] | None = None


_HEALTH_RECORD_CREATE_PAYLOAD_MODELS = {
    "observations": HealthRecordObservationDraft,
    "conditions": HealthRecordConditionDraft,
    "medications": HealthRecordMedicationDraft,
    "encounters": HealthRecordEncounterDraft,
}
_HEALTH_RECORD_UPDATE_PAYLOAD_MODELS = {
    "observations": ObservationUpdate,
    "conditions": ConditionUpdate,
    "medications": MedicationUpdate,
    "encounters": EncounterUpdate,
}


class HealthRecordAction(BaseModel):
    action: HealthRecordActionType
    resource: HealthRecordActionResource
    target_member_id: str
    record_id: str | None = None
    payload: dict[str, Any] | None = None

    @field_validator("target_member_id", "record_id")
    @classmethod
    def validate_text_fields(cls, value: str | None) -> str | None:
        if value is None:
            return value
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field is required.")
        return cleaned

    @model_validator(mode="after")
    def validate_action_payload(self) -> "HealthRecordAction":
        if self.action == "create":
            if self.record_id is not None:
                raise ValueError("record_id is not allowed for create actions.")
            if self.payload is None:
                raise ValueError("payload is required for create actions.")
            payload_model = _HEALTH_RECORD_CREATE_PAYLOAD_MODELS[self.resource].model_validate(self.payload)
            self.payload = payload_model.model_dump(exclude_none=True)
            return self

        if self.record_id is None:
            raise ValueError("record_id is required for update and delete actions.")

        if self.action == "delete":
            if self.payload not in (None, {}):
                raise ValueError("payload is not allowed for delete actions.")
            self.payload = None
            return self

        if self.payload is None:
            raise ValueError("payload is required for update actions.")

        payload_model = _HEALTH_RECORD_UPDATE_PAYLOAD_MODELS[self.resource].model_validate(self.payload)
        dumped_payload = payload_model.model_dump(exclude_unset=True)
        if not dumped_payload:
            raise ValueError("payload must include at least one field for update actions.")
        self.payload = dumped_payload
        return self


class HealthRecordDraft(BaseModel):
    summary: str = ""
    actions: list[HealthRecordAction] = Field(default_factory=list)

File: frontend/src/api/health.ts
import type { AuthSession } from "../auth/session";

import { deleteAuthorized, getAuthorized, sendAuthorized } from "./http";


export type HealthSummaryRecord = {
  id: string;
  member_id: string;
  category: string;
  label: string;
  value: string;
  status: "good" | "warning" | "alert";
  generated_at: string;
  created_at: string;
};

export type CarePlanRecord = {
  id: string;
  member_id: string;
  assignee_member_id: string | null;
  category: string;
  icon_key: "medication" | "exercise" | "checkup" | "meal" | "rest" | "social" | "general" | null;
  time_slot: "清晨" | "上午" | "午后" | "晚间" | "睡前" | null;
  title: string;
  description: string;
  notes: string | null;
  status: "active" | "completed" | "cancelled";
  scheduled_at: string | null;
  completed_at: string | null;
  generated_by: "ai" | "manual";
  created_at: string;
  updated_at: string;
};

export type DashboardReminder = CarePlanRecord & {
  member_name: string;
};

export type DashboardReminderGroup = {
  time_slot: "清晨" | "上午" | "午后" | "晚间" | "睡前";
  reminders: DashboardReminder[];
};

export type ObservationRecord = {
  id: string;
  member_id: string;
  category: "chronic-vitals" | "lifestyle" | "body-vitals";
  code: string;
  display_name: string;
  value: number | null;
  value_string: string | null;
  unit: string | null;
  context: string | null;
  effective_at: string;
  source: "manual" | "device" | "ai-extract";
  device_name: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type ConditionRecord = {
  id: string;
  member_id: string;
  category: "diagnosis" | "chronic" | "allergy" | "family-history";
  display_name: string;
  clinical_status: "active" | "inactive" | "resolved";
  onset_date: string | null;
  source: "manual" | "ai-extract";
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type MedicationRecord = {
  id: string;
  member_id: string;
  name: string;
  indication: string | null;
  dosage_description: string | null;
  status: "active" | "stopped";
  start_date: string | null;
  end_date: string | null;
  source: "manual" | "ai-extract";
  created_at: string;
  updated_at: string;
};

export type EncounterRecord = {
  id: string;
  member_id: string;
  type: "outpatient" | "inpatient" | "checkup" | "emergency";
  facility: string | null;
  department: string | null;
  attending_physician: string | null;
  date: string;
  summary: string | null;
  source: "manual" | "ai-extract";
  created_at: string;
  updated_at: string;
};

export type SleepRecord = {
  id: string;
  member_id: string;
  start_at: string;
  end_at: string;
  total_minutes: number;
  deep_minutes: number | null;
  rem_minutes: number | null;
  light_minutes: number | null;
  awake_minutes: number | null;
  efficiency_score: number | null;
  is_nap: boolean;
  source: "manual" | "device";
  device_name: string | null;
  created_at: string;
};

export type WorkoutRecord = {
  id: string;
  member_id: string;
  type: string;
  start_at: string;
  end_at: string;
  duration_minutes: number;
  energy_burned: number | null;
  distance_meters: number | null;
  avg_heart_rate: number | null;
  source: "manual" | "device";
  device_name: string | null;
  notes: string | null;
  created_at: string;
};

export type DashboardMemberSummary = {
  member: {
    id: string;
    name: string;
    gender: string;
    avatar_url: string | null;
    blood_type: string | null;
  };
  health_summaries: HealthSummaryRecord[];
};

export type DashboardResponse = {
  members: DashboardMemberSummary[];
  today_reminders: DashboardReminder[];
  reminder_groups?: DashboardReminderGroup[];
};

export function getDashboard(session: AuthSession) {
  return getAuthorized<DashboardResponse>("/api/dashboard", session);
}

export function listObservations(session: AuthSession, memberId: string) {
  return getAuthorized<ObservationRecord[]>(`/api/members/${memberId}/observations`, session);
}

export function listConditions(session: AuthSession, memberId: string) {
  return getAuthorized<ConditionRecord[]>(`/api/members/${memberId}/conditions`, session);
}

export function listMedications(session: AuthSession, memberId: string) {
  return getAuthorized<MedicationRecord[]>(`/api/members/${memberId}/medications`, session);
}

export function listEncounters(session: AuthSession, memberId: string) {
  return getAuthorized<EncounterRecord[]>(`/api/members/${memberId}/encounters`, session);
}

export function listCarePlans(session: AuthSession, memberId: string) {
  return getAuthorized<CarePlanRecord[]>(`/api/members/${memberId}/care-plans`, session);
}

export function listSleepRecords(session: AuthSession, memberId: string) {
  return getAuthorized<SleepRecord[]>(`/api/members/${memberId}/sleep-records`, session);
}

export function listWorkoutRecords(session: AuthSession, memberId: string) {
  return getAuthorized<WorkoutRecord[]>(`/api/members/${memberId}/workout-records`, session);
}

export function listHealthSummaries(session: AuthSession, memberId: string) {
  return getAuthorized<HealthSummaryRecord[]>(`/api/members/${memberId}/health-summaries`, session);
}

// ─── Observation CRUD ─────────────────────────────────────────────────────────

export type ObservationCreatePayload = {
  category: "chronic-vitals" | "lifestyle" | "body-vitals";
  code: string;
  display_name: string;
  value?: number | null;
  value_string?: string | null;
  unit?: string | null;
  context?: string | null;
  effective_at: string;
  source?: "manual" | "device" | "ai-extract";
  device_name?: string | null;
  notes?: string | null;
};

export type ObservationUpdatePayload = Partial<ObservationCreatePayload>;

export function createObservation(session: AuthSession, memberId: string, payload: ObservationCreatePayload) {
  return sendAuthorized<ObservationRecord, ObservationCreatePayload>(
    `/api/members/${memberId}/observations`,
    session,
    { method: "POST", payload },
  );
}

export function updateObservation(session: AuthSession, memberId: string, recordId: string, payload: ObservationUpdatePayload) {
  return sendAuthorized<ObservationRecord, ObservationUpdatePayload>(
    `/api/members/${memberId}/observations/${recordId}`,
    session,
    { method: "PUT", payload },
  );
}

export function deleteObservation(session: AuthSession, memberId: string, recordId: string) {
  return deleteAuthorized(`/api/members/${memberId}/observations/${recordId}`, session);
}

// ─── Condition CRUD ───────────────────────────────────────────────────────────

export type ConditionCreatePayload = {
  category: "diagnosis" | "chronic" | "allergy" | "family-history";
  display_name: string;
  clinical_status?: "active" | "inactive" | "resolved";
  onset_date?: string | null;
  source?: "manual" | "ai-extract";
  notes?: string | null;
};

export type ConditionUpdatePayload = Partial<ConditionCreatePayload>;

export function createCondition(session: AuthSession, memberId: string, payload: ConditionCreatePayload) {
  return sendAuthorized<ConditionRecord, ConditionCreatePayload>(
    `/api/members/${memberId}/conditions`,
    session,
    { method: "POST", payload },
  );
}

export function updateCondition(session: AuthSession, memberId: string, recordId: string, payload: ConditionUpdatePayload) {
  return sendAuthorized<ConditionRecord, ConditionUpdatePayload>(
    `/api/members/${memberId}/conditions/${recordId}`,
    session,
    { method: "PUT", payload },
  );
}

export function deleteCondition(session: AuthSession, memberId: string, recordId: string) {
  return deleteAuthorized(`/api/members/${memberId}/conditions/${recordId}`, session);
}

// ─── Medication CRUD ──────────────────────────────────────────────────────────

export type MedicationCreatePayload = {
  name: string;
  indication?: string | null;
  dosage_description?: string | null;
  status?: "active" | "stopped";
  start_date?: string | null;
  end_date?: string | null;
  source?: "manual" | "ai-extract";
};

export type MedicationUpdatePayload = Partial<MedicationCreatePayload>;

export function createMedication(session: AuthSession, memberId: string, payload: MedicationCreatePayload) {
  return sendAuthorized<MedicationRecord, MedicationCreatePayload>(
    `/api/members/${memberId}/medications`,
    session,
    { method: "POST", payload },
  );
}

export function updateMedication(session: AuthSession, memberId: string, recordId: string, payload: MedicationUpdatePayload) {
  return sendAuthorized<MedicationRecord, MedicationUpdatePayload>(
    `/api/members/${memberId}/medications/${recordId}`,
    session,
    { method: "PUT", payload },
  );
}

export function deleteMedication(session: AuthSession, memberId: string, recordId: string) {
  return deleteAuthorized(`/api/members/${memberId}/medications/${recordId}`, session);
}

// ─── Encounter CRUD ───────────────────────────────────────────────────────────

export type EncounterCreatePayload = {
  type: "outpatient" | "inpatient" | "checkup" | "emergency";
  facility?: string | null;
  department?: string | null;
  attending_physician?: string | null;
  date: string;
  summary?: string | null;
  source?: "manual" | "ai-extract";
};

export type EncounterUpdatePayload = Partial<EncounterCreatePayload>;

export function createEncounter(session: AuthSession, memberId: string, payload: EncounterCreatePayload) {
  return sendAuthorized<EncounterRecord, EncounterCreatePayload>(
    `/api/members/${memberId}/encounters`,
    session,
    { method: "POST", payload },
  );
}

export function updateEncounter(session: AuthSession, memberId: string, recordId: string, payload: EncounterUpdatePayload) {
  return sendAuthorized<EncounterRecord, EncounterUpdatePayload>(
    `/api/members/${memberId}/encounters/${recordId}`,
    session,
    { method: "PUT", payload },
  );
}

export function deleteEncounter(session: AuthSession, memberId: string, recordId: string) {
  return deleteAuthorized(`/api/members/${memberId}/encounters/${recordId}`, session);
}

// ─── SleepRecord CRUD ─────────────────────────────────────────────────────────

export type SleepRecordCreatePayload = {
  start_at: string;
  end_at: string;
  total_minutes: number;
  deep_minutes?: number | null;
  rem_minutes?: number | null;
  light_minutes?: number | null;
  awake_minutes?: number | null;
  efficiency_score?: number | null;
  is_nap?: boolean;
  source?: "manual" | "device";
  device_name?: string | null;
};

export type SleepRecordUpdatePayload = Partial<SleepRecordCreatePayload>;

export function createSleepRecord(session: AuthSession, memberId: string, payload: SleepRecordCreatePayload) {
  return sendAuthorized<SleepRecord, SleepRecordCreatePayload>(
    `/api/members/${memberId}/sleep-records`,
    session,
    { method: "POST", payload },
  );
}

export function updateSleepRecord(session: AuthSession, memberId: string, recordId: string, payload: SleepRecordUpdatePayload) {
  return sendAuthorized<SleepRecord, SleepRecordUpdatePayload>(
    `/api/members/${memberId}/sleep-records/${recordId}`,
    session,
    { method: "PUT", payload },
  );
}

export function deleteSleepRecord(session: AuthSession, memberId: string, recordId: string) {
  return deleteAuthorized(`/api/members/${memberId}/sleep-records/${recordId}`, session);
}

// ─── WorkoutRecord CRUD ───────────────────────────────────────────────────────

export type WorkoutRecordCreatePayload = {
  type: string;
  start_at: string;
  end_at: string;
  duration_minutes: number;
  energy_burned?: number | null;
  distance_meters?: number | null;
  avg_heart_rate?: number | null;
  source?: "manual" | "device";
  device_name?: string | null;
  notes?: string | null;
};

export type WorkoutRecordUpdatePayload = Partial<WorkoutRecordCreatePayload>;

export function createWorkoutRecord(session: AuthSession, memberId: string, payload: WorkoutRecordCreatePayload) {
  return sendAuthorized<WorkoutRecord, WorkoutRecordCreatePayload>(
    `/api/members/${memberId}/workout-records`,
    session,
    { method: "POST", payload },
  );
}

export function updateWorkoutRecord(session: AuthSession, memberId: string, recordId: string, payload: WorkoutRecordUpdatePayload) {
  return sendAuthorized<WorkoutRecord, WorkoutRecordUpdatePayload>(
    `/api/members/${memberId}/workout-records/${recordId}`,
    session,
    { method: "PUT", payload },
  );
}

export function deleteWorkoutRecord(session: AuthSession, memberId: string, recordId: string) {
  return deleteAuthorized(`/api/members/${memberId}/workout-records/${recordId}`, session);
}
